### ISSUE TITLE
two issues in strtol (strtoll too)

### ISSUE THREAD

**Issue by parisseb** [CE-Programming/toolchain#514]
**First issue:**
When overflow is detected, the string should still be parsed up to the end to detect invalid characters in this basis.
For example, in `strtol` for 1234567890123 there is no invalid char, `*endptr` should be 0.

**Second issue:**
The overflow test is wrong. If 10x>=2^31, there is no reason to have 10x as a long be smaller than x.
Example: parse 5783484780, at the last Horner step, x=578348478, 10x is >= 2^31, but it's representation as a 32 bits signed integer is 1488517484, and that's greater than x. 
A safe way to check overflow: compare sum to (2^31-1)/radix (should perhaps be precomputed for speed in basis 10) before doing sum*=radix. Cf. e.g. https://stackoverflow.com/questions/7457163/what-is-the-implementation-of-strtol

**Comment by mateoconlechuga** [comment]
This is not a suitable bug report. Please provide code with expected input/output.

**Comment by parisseb** [comment]
?? The description is clear enough, and I gave a link with working code that you just have to copy/paste, I just checked it inside KhiCAS.

**Comment by adriweb** [comment]
Some example code (use `dbg_printf` for the CE...) and expected output:

```c
#include <stdio.h>
#include <stdlib.h>

int main(void)
{
    const char* str = "1234567890123";
    char* endptr = NULL;
    long ret = strtol(str, &endptr, 10);
    printf("ret = %ld, endptr = %p, str at %p, endptr-str = %ld\n", ret, endptr, str, endptr-str);
    return 0;
}
```

Sample good output:
```
ret = 1234567890123, endptr = 0x10048bf71, str at 0x10048bf64, endptr-str = 13
```

**Comment by adriweb** [comment]
On the CE it produces this wrong output:

```
ret = 2147483647, endptr = 0xd1cac6, str at 0xd1cabc, endptr-str = 10
```

**Comment by adriweb** [comment]
A corrected code for the CE would be to use the `long long` variant @parisseb :
```c
long long ret = strtoll(str, &endptr, 10);
```

In which case, we get the correct output:
```
ret = 1234567890123, endptr = 0xd1cad4, str at 0xd1cac7, endptr-str = 13
```

**Comment by parisseb** [comment]
This is not a valid fix because the same issues will occur for strtoll, just for larger inputs (that was my first workaround for KhiCAS, then I realized it's not a real fix).

**Comment by adriweb** [comment]
Yes, correct, this just pushes the issue back further.

(As also mentioned by @calc84maniac on Discord:)
> **calc84maniac — _Today at 14:56_**
> yeah using long long is just causing the problem to happen with even more characters, assuming that routine has the same issue
> and if you want 32-bit output then you don't get the desired saturation

**Comment by adriweb** [comment]
Not sure how valid this is, really, but if, on the computer, we force the use of an `int32_t` (like a CE `long), we get this:

```
ret = 1912276171, endptr = 0x10470ff65, *endptr = 0, str at 0x10470ff58, endptr-str = 13
```

Which I guess is more correct than on the CE.

**Comment by calc84maniac** [comment]
What exactly do you mean by "forcing the use of an `int32_t`"? Assigning the output of `strtol` to it? Because that doesn't limit the operation of `strtol` to 32 bits, it simply truncates the output. You would have to test on a system with 32-bit `long` (e.g. 32-bit Unix or any version of Windows) to see what the expected output on a CE should be.

**Comment by adriweb** [comment]
Well I tested on godbolt with `-m32` (is that good enough?)

And I suppose I actually meant to paste _this_ output: (still a correct `*endptr` and ptr difference which is what I was really looking at anyway)

`ret = 1912276171, endptr = 0x402015, *endptr = 0, str at 0x402008, endptr-str = 13`

**Comment by adriweb** [comment]
Fixed by [cb7d415](https://github.com/CE-Programming/toolchain/commit/cb7d41562b950f6f95ccc82fed4a75fc58f64dae) :)
